### PR TITLE
feature/CONTROLLER Update controller

### PR DESCRIPTION
**Pull request by ichabanivan**
- Added DEBUG Mode
- Added types for selector
- Added types for actions
- Added types for subscribe
- Update README
- Update reducer store data to next model { connected: boolean; actual: I; initial: I; }

- Remove named export for controller
- Add checking for all input data
- Add utils (isGeneratorFunction, isString, isBoolean, isArray, isObject)
- Simplify record for useEffect in useReducerSubscribe and useSubscribe
- Update documentation
- Add comments

- Create example
- Update README.md (Changed examples, Describe API)
- Update npm ignore
- Create todo.md
- Update tsconfig
- Create different actions (all,actual,initial,connected)

- Remove extra types in the controller
- Change logs to different view
- Create  common selector for CSD

### CODE REVIEW

**Comment by ichabanivan** [comment]
image

File: example/store.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+// NOTE Build the middleware to run our Saga
+const sagaMiddleware = createSagaMiddleware();
+
+// NOTE Create store outside of root to be able dispatch actions from anywhere!
+const store = createStore(
+  combineReducers({
+    // ...your other reducers here
+    // you have to pass controllerReducer under 'controller' key,
+    controller: controllerReducer,
```
Comment (sajera) [inline example/store.ts:15]: as addition ability we may implement method to provide custom redux name for example
```javascript

...controllerReducerCustom('myCustomReducrName'); //  => { [name]:  controllerReducer }
```

File: src/controller.ts [hunk 2]
```diff
@@ -3,71 +3,108 @@
...
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    subscriber: any,
   }) {
-    // TODO Add validation for all properties
-    this.name = uniqueId(prefix);
-    this.initial = initial;
-    this.subscriber = subscriber;
-    // TODO forbid to pass properties
+    // NOTE Setup DEBUG value if present and is valid
+    this.DEBUG = isBoolean(DEBUG) ? DEBUG : false;
```
Comment (sajera) [inline src/controller.ts:52]: use local storage to enable debug mode

File: src/utils.ts [hunk 3]
```diff
@@ -15,15 +15,32 @@ const idCounter: Record<string, number> = {};
  * uniqueId()
  * // => '105'
  */
-export const uniqueId = (prefix = '$controller$') => {
-  if (!idCounter[prefix]) {
-    idCounter[prefix] = 0;
+export const uniqueId = (prefix = 'controller_') => {
+  if (!counter[prefix]) {
```
Comment (sajera) [inline src/utils.ts:19]: should generate not only unique but new one each time

**Review by sajera — CHANGES_REQUESTED** [review]

**Pull request merged**